### PR TITLE
enable ssl client connections to brokers (#334)

### PR DESCRIPTION
**Pull request by CharlesWall**
Implement ssl socket for client/broker connections. only works with +0.9

### CODE REVIEW

File: lib/client.js [hunk 1]
```diff
@@ -60,20 +64,47 @@ var Client = function (connectionString, clientId, zkOptions, noAckBatchOptions)
...
-                var broker = brokers[key];
-                var addr = broker.host + ':' + broker.port;
-                self.brokers[addr] = self.createBroker(broker.host, broker.port);
+            .forEach(function (key, index) {
+                var brokerProfile = brokers[key];
+                var addr = brokerProfile.host + ':' + brokerProfile.port;
+                self.brokerProfiles[addr] = brokerProfile;
+                if (self.ssl) {
+                    var sslAddress = brokerProfile.endpoints.find(function(endpoint) {
+                        return endpoint.startsWith('SSL://');
```
Comment (hyperlink) [inline lib/client.js]: this won't work in `0.12.14` and below.

File: lib/client.js [hunk 2]
```diff
@@ -60,20 +64,47 @@ var Client = function (connectionString, clientId, zkOptions, noAckBatchOptions)
...
+                    });
+                    if(!sslAddress) {
+                        var err = new Error('No SSL kafka endpoint found');
+                        self.emit('error', err);
+                    }
+                    sslAddress = splitOnFirst(sslAddress, '://')[1];
```
Comment (hyperlink) [inline lib/client.js]: Couldn't you use the `url` module to do the parsing?

File: test/test.consumer.js [hunk 3]
```diff
@@ -29,368 +30,396 @@ function offsetOutOfRange (topic, consumer) {
...
+    var suiteTimeout = testParameters.suiteTimeout || null;
+    var suiteName = testParameters.name;
+
+
+    function createClient() {
+        var clientId = 'kafka-node-client-' + uuid.v4();
+        return new Client(host, clientId, undefined, undefined, sslOptions);
+    }
+
+    describe(suiteName, function () {
```
Comment (CharlesWall) [inline test/test.consumer.js]: everything after this is just indenting to put it inside of the forEach above.

File: docker/certs/ca-key [hunk 4]
```diff
@@ -0,0 +1,18 @@
+-----BEGIN RSA PRIVATE KEY-----
```
Comment (CharlesWall) [inline docker/certs/ca-key]: it feels wrong to commit up certs and a private key, but they are only for development use.

File: docker-compose.yml [hunk 5]
```diff
@@ -8,10 +8,19 @@ services:
...
     ports:
       - "9092:9092"
+      - "9093:9093"
     depends_on:
       - zookeeper
     environment:
       KAFKA_ADVERTISED_HOST_NAME: ${KAFKA_ADVERTISED_HOST_NAME}
       KAFKA_ZOOKEEPER_CONNECT: zookeeper:2181
+      KAFKA_LISTENERS: "PLAINTEXT://:9092,SSL://:9093"
+      KAFKA_ADVERTISED_LISTENERS: "PLAINTEXT://192.168.64.2:9092,SSL://192.168.64.2:9093"
```
Comment (hyperlink) [inline docker-compose.yml]: you will need to change the hardcoded IP to `${KAFKA_ADVERTISED_HOST_NAME}`

File: lib/client.js [hunk 6]
```diff
@@ -509,8 +535,23 @@ Client.prototype.setupBroker = function (host, port, longpolling, brokers) {
...
+
     var self = this;
-    var socket = net.createConnection(port, host);
+    var socket;
+    if (self.ssl) {
+        socket = tls.connect(port, host, self.sslOptions);
+    } else {
+        socket = net.createConnection(port, host);
+    }
     socket.addr = host + ':' + port;
```
Comment (hyperlink) [inline lib/client.js]: Please use `addr` here so we don't have to concat `host` and `port` again.

File: README.md [hunk 7]
```diff
@@ -16,11 +16,12 @@ Follow the [instructions](http://kafka.apache.org/documentation.html#quickstart)
 
 # API
 ## Client
-### Client(connectionString, clientId, [zkOptions], [noAckBatchOptions])
+### Client(connectionString, clientId, [zkOptions], [noAckBatchOptions], [sslOptions])
 * `connectionString`: Zookeeper connection string, default `localhost:2181/`
 * `clientId`: This is a user-supplied identifier for the client application, default `kafka-node-client`
 * `zkOptions`: **Object**, Zookeeper options, see [node-zookeeper-client](https://github.com/alexguan/node-zookeeper-client#client-createclientconnectionstring-options)
 * `noAckBatchOptions`: **Object**, when requireAcks is disabled on Producer side we can define the batch properties, 'noAckBatchSize' in bytes and 'noAckBatchAge' in milliseconds. The default value is `{ noAckBatchSize: null, noAckBatchAge: null }` and it acts as if there was no batch
+* `sslOptions`: **Object**, options to be passed to the tls broker sockets, ex. { rejectUnauthorized: false }
```
Comment (hyperlink) [inline README.md]: can we add a blurb about how this is only supported for kafka 0.9 and above?

File: lib/client.js (hunk 6 above)
Comment (CharlesWall) [inline lib/client.js]: `addr` may have changed because of :545/:546 it is important that we recalculate it here.

File: docker-compose.yml (hunk 5 above)
Comment (CharlesWall) [inline docker-compose.yml]: will fix.

File: lib/client.js (hunk 6 above)
Comment (hyperlink) [inline lib/client.js]: yup, I see that now. 👍

**Comment by hyperlink** [comment]
Feedback on this PR @haio or @estliberitas ?

**Comment by hyperlink** [comment]
Thank you for your contribution @CharlesWall !

**Pull request merged**